### DISCUSSION TITLE
Make `FluentTheme` instance with changed color palette, dynamically

### DISCUSSION THREAD

**Discussion by AzuxirenLeadGuy** [AvaloniaUI/Avalonia#12042] [Q&A] [answered]
I wish to have multiple themes (All `FluentTheme` instances but with different color palettes) supported in my app. These themes would be made dynamically and and be applied in C#. How to do this in v11.0.0 ? 

From what I understand (might be incorrect, feel free to correct me), maybe it would begin like 

```csharp
void ApplyTheme(App currentApplication, ThemeVariant themeType)
{
	currentApplication.Styles.Clear(); // Remove previously present style 
	var theme = new FluentTheme()
	{
		Palettes = 
		{
			// Apply changes in color, how to assign?
		}
	};
	currentApplication.Styles.Add(theme);
	currentApplication.RequestedThemeVariant = themeType;
}
```

Also, congratulations for the latest release

**Comment by timunie** [comment]
the common way via DynamicResources should still work. Just override the brushes / colors you whish. 

```xml
App.Current.Resources["MyBrush"] = new SolidColorBrush(Colors.Blue);
```

**Reply by AzuxirenLeadGuy** [reply]
Thanks for your reply. Just to be clear, If I wish to use a palette obtained from https://theme.xaml.live/ ,  something like
```xml
<FluentTheme>
  <FluentTheme.Palettes>
    <ColorPaletteResources x:Key="Light" 
    Accent="#ff8961cc" 
    AltHigh="White" 
    AltLow="White"  ...

```

I should equivalently do


```csharp
App.Current.Resources["Accent"] = new SolidColorBrush(Color.FromRgb(0x89, 0x61, 0xcc)); // 0xff8961cc 
App.Current.Resources["AltHigh"] = new SolidColorBrush(Color.FromRgb(0xff, 0xff, 0xff)); //  White 
App.Current.Resources["AltLow"] = new SolidColorBrush(Color.FromRgb(0xff, 0xff, 0xff)); //  White
```

Is that right?

**Reply by timunie** [reply]
hmmm ... looks like we have this: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/ColorPaletteResources.Properties.cs 

So you need to override the given Keys (`AltLow -> SytemAltLowColor` etc), not the name of the property. Or you can do this in code behind:

In App.xaml.cs:
```cs
// your access to global palette
public static ColorPaletteResources Accents { get; } = new ColorPaletteResources()
{
    Accent = Colors.Red
};

public override void OnFrameworkInitializationCompleted()
{
    // Attach our palette to both themes, you can also have different for light and dark
    (Styles[0] as FluentTheme).Palettes.Add(ThemeVariant.Light, Accents);
    (Styles[0] as FluentTheme).Palettes.Add(ThemeVariant.Dark, Accents);
...
```

And then in your ViewModel or from ColorPicker:

```cs
private void ColorView_OnColorChanged(object? sender, ColorChangedEventArgs e)
{
    App.Accents.Accent = e.NewColor;
}
```

works for me in a test app. The benefit is, the shades are calculated for you.

**Comment by maxkatz6** [comment, marked as answer]
You cannot change a palette dynamically, but you can reassign a whole FluentTheme with different palettes.
So, from your code:
```
void ApplyTheme(App currentApplication, ThemeVariant themeType)
{
	currentApplication.Styles.Clear(); // Remove previously present style 
	var theme = new FluentTheme()
	{
		Palettes = 
		{
			[ThemeVariant.Light] = new ColorPaletteResources() { ... },
			[ThemeVariant.Dark] = new ColorPaletteResources() { ... },
		}
	};
	currentApplication.Styles.Add(theme);
	currentApplication.RequestedThemeVariant = themeType;
}
```

That if you need multiple instances of the same theme and different values for Dark/Light palettes, of course. If you just need two variants - light and dark with your colors, you don't need that. See https://docs.avaloniaui.net/docs/next/basics/user-interface/styling/themes/fluent#creating-custom-color-palettes

**Reply by maxkatz6** [reply]
And if you just want to create FluentTheme once, but generate ColorPaletteResources from the code, you can follow @timunie example with OnFrameworkInitializationCompleted. Just keep in mind, that palettes are static and you can't change them after any control was created (it won't be applied).

**Reply by AzuxirenLeadGuy** [reply]
This was exactly what I was looking for. Thank you so much, and sorry for my late reply.

**Reply by AzuxirenLeadGuy** [reply]
Okay... a bit of an issue...

I have a ComboBox in the page. After the theme is updated using the above method, If I click on the comboBox, it crashes with the exception

```
Unhandled exception. System.InvalidOperationException: The control already has a visual parent.
   at Avalonia.Visual.ValidateVisualChild(Visual c) in /_/src/Avalonia.Base/Visual.cs:line 659
   at Avalonia.Collections.AvaloniaList`1.Add(T item) in /_/src/Avalonia.Base/Collections/AvaloniaList.cs:line 205
   at Avalonia.Controls.Presenters.ContentPresenter.UpdateChild(Object content) in /_/src/Avalonia.Controls/Presenters/ContentPresenter.cs:line 484
   at Avalonia.Controls.Presenters.ContentPresenter.ContentChanged(AvaloniaPropertyChangedEventArgs e) in /_/src/Avalonia.Controls/Presenters/ContentPresenter.cs:line 677
   at Avalonia.Controls.Presenters.ScrollContentPresenter.OnPropertyChanged(AvaloniaPropertyChangedEventArgs change) in /_/src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs:line 693
   at Avalonia.Animation.Animatable.OnPropertyChangedCore(AvaloniaPropertyChangedEventArgs change) in /_/src/Avalonia.Base/Animation/Animatable.cs:line 106
......
```

What am I doing wrong here?

**Reply by timunie** [reply]
most likely you try to render the same control in several parents. Often it means you do not use a DataTemplate where one is needed. But without having a minimal sample, I cannot help any further.

**Reply by AzuxirenLeadGuy** [reply]
Thank you for your reply @timunie . I am indeed not using a DataTemplate, and I am not sure what it is, but I'll check it out in the docs, so thank you for guiding me there. 

Right now, as a workaround, I am reloading the ComboBox (by creating a new instance of it and replacing the old ComboBox instance in the panel) and now it works fine. So somehow, reloading the ComboBox is what is working for me. There are other controls in the page as well, like a `Slider`, `Button`, `Expander` and `TextBlock`, but it is only the `ComboBox` instance that throws the "control already has a visual parent" exception (when not reloaded). 

I will make a minimal sample and show it to you, but it will take a while

**Reply by AzuxirenLeadGuy** [reply]
I have uploaded a minimal example here : https://github.com/AzuxirenLeadGuy/avalonia-test-app

The logic for changing the theme is in `ava_prac/App.cs`, while the app contains only a single view, defined in `ava_prac/Views/MainView.cs`

As you can see, there are multiple controls including with a combo box that contains a list of available themes, and a button which applies the said theme (checks the index of the combo box, and calls the function to apply the theme which is defined in App.cs).

When the application is running (I am running the desktop app, on a linux machine), after applying a theme, if the combobox is interacted with, the application crashes. There are multiple other controls in the application as well which are not affected at all, except for yet another ComboBox inside the expander. If this inner ComboBox is assigned any value before changing the theme, and then interacted again after changing the theme, it throws the same exception. I believe this is a bug with ComboBox.

The exception thrown in both the cases is the same, as shown 

```
Unhandled exception. System.InvalidOperationException: The control already has a visual parent.
```

**Reply by timunie** [reply]
woah I'm not used to do such things in code behind only, sorry. Consider to use the MVVM concept. It has some steep learning curve, but helps in long term.

Anyway, I found the issue doesn't happen if I use the proposal I wrote earlier. Seems only to happen if one swaps the entire theme out and in again. Maybe the order of it is important, but I have no idea tbh.

**Reply by schnerring** [reply]
@maxkatz6 Is it possible to combine `ThemedDictionaries` with `ColorPaletteResources`? Using `DynamicResource` doesn't work. How do I approach this?

**Reply by maxkatz6** [reply]
@schnerring what exactly are you trying to achieve? I don't understand a question about combining ThemedDictionaries with ColorPaletteResources. DynamicResource should work fine, if you have to change theme or change accent color.

See https://docs.avaloniaui.net/docs/next/basics/user-interface/styling/themes/fluent#creating-custom-color-palettes
Also, note, properties other than Accent color are not bindable.

**Reply by schnerring** [reply]
> Also, note, properties other than Accent color are not bindable.

That answers my question. Sorry for overlooking that, it's mentioned in the docs.

I have defined a light and dark color palette inside `ThemeDictionaries`. What I tried to do is override properties other than `Accent` of the fluent theme by using `DynamicResource`s (which qualifies as binding, right?)

**Reply by timunie** [reply]
Yeah, DynamicResources are kind of "Binding" technically and will change on the fly if changed.
